internal/pkg/term/syncbuffer/syncbuffer.go

```diff
@@ -0,0 +1,149 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a synchronized buffer to store and access logs from multiple goroutines.
+package syncbuffer
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+	"sync"
+
+	"golang.org/x/term"
+)
+
+// maxLogLines is the maximum number of lines to display in the terminal.
+const maxLogLines = 5
+
+// fileWriter is the interface to write to a file.
+type fileWriter interface {
+	io.Writer
+}
+
+// SyncBuffer is a synchronized buffer that can be used to store output data and coordinate between multiple goroutines.
+type SyncBuffer struct {
+	BufMu sync.Mutex   // BufMu is a mutex that can be used to protect access to the buffer.
+	Buf   bytes.Buffer // Buf is the buffer that stores the data.
+
+	Done chan struct{} // Done is a channel that can be used to signal when the operations are complete.
+}
+
+// Write appends the given bytes to the buffer.
+func (b *SyncBuffer) Write(p []byte) (n int, err error) {
+	b.BufMu.Lock()
+	defer b.BufMu.Unlock()
+	return b.Buf.Write(p)
+}
+
+// Strings returns an empty slice if the buffer is empty.
+// Otherwise, it returns a slice of all the lines stored in the buffer.
+func (b *SyncBuffer) strings() []string {
+	b.BufMu.Lock()
+	defer b.BufMu.Unlock()
+
+	// Split the buffer bytes into lines.
+	lines := strings.Split(strings.TrimSpace(b.Buf.String()), "\n")
+
+	return lines
+}
+
+// IsDone returns true if the Done channel has been closed, otherwise return false.
+func (b *SyncBuffer) IsDone() bool {
+	select {
+	case <-b.Done:
+		return true
+	default:
+		return false
+	}
+}
+
+// MarkDone closes the Done channel.
+func (b *SyncBuffer) MarkDone() {
+	close(b.Done)
+}
+
+// TermPrinter is a printer to display logs in the terminal.
+type TermPrinter struct {
+	Term             fileWriter
+	Buf              *SyncBuffer
+	PrevWrittenLines int
+	TermWidth        int
+}
+
+// NewTermPrinter returns a new instance of TermPrinter that writes logs to the given file writer and reads logs from a new synchronized buffer.
+func NewTermPrinter(fw fileWriter) *TermPrinter {
```

Comment:
- Can we expose `fileWriter` since it's an input to the constructor?
- To make the struct flexible, consider taking as input the "number of lines" to print from the buffer. This way we don't have to limit ourselves to 5 
- Should the `termWidth` field be calculated here in the constructor? this way clients don't have to call the `TerminalWidth` method and pass the value down to the object